### ISSUE TITLE
r/aws_s3_bucket_object: `etag` can’t be used with any server encryption, but docs mention a different thing

### ISSUE THREAD

**Issue by iamakulov** [terraform-providers/terraform-provider-aws#5033]
Currently, [docs](https://www.terraform.io/docs/providers/aws/r/s3_bucket_object.html) state that `etag` in `aws_s3_bucket_object` can’t be used only with KMS encryption:

> `etag` - (Optional) Used to trigger updates. The only meaningful value is `${md5(file("path/to/file"))}`. This attribute is not compatible with KMS encryption, `kms_key_id` or `server_side_encryption = "aws:kms"`.

However, when I try to use it with AES256 encryption, I receive the

```
Error: module.chorus-backend.aws_s3_bucket_object.env-file: "etag": conflicts with server_side_encryption
```

error.

https://github.com/terraform-providers/terraform-provider-aws/issues/2244 brought this issue about docs. https://github.com/terraform-providers/terraform-provider-aws/pull/4928 closed it – however, the docs still talk only about KMS encryption. Looks like it’s either the issue in docs again (Terraform is right, docs are wrong) or in the Terraform itself (docs list correct expectations, but Terraform implements them wrong).

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

v0.11.7

### Affected Resource(s)

* `aws_s3_bucket_object`

### Terraform Configuration Files

```hcl
resource "aws_s3_bucket_object" "env-file" {
  key                    = "${var.env-name}/.env"
  bucket                 = "${var.chorus-configs-bucket-name}"
  content                = "${data.template_file.env-file.rendered}"
  server_side_encryption = "AES256"
  etag                   = "${md5("${data.template_file.env-file.rendered}")}"
}

data "template_file" "env-file" {
  template = "${file("${path.module}/.env")}"

  vars {
    app-domain = "${var.app-domain}"
    aws-region = "${var.aws-region}"
  }
}
```

### Debug Output

https://gist.github.com/iamakulov/7525920a9ebeaa0ca5af2b465a4be878

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

I’m using Windows.

### References

* https://github.com/terraform-providers/terraform-provider-aws/issues/2244
* https://github.com/terraform-providers/terraform-provider-aws/pull/4928

**Comment by marek-obuchowicz** [comment]
Same issue happens on mac os x

**Comment by nwalke** [comment]
If you look at the AWS docs, I think the Terraform docs are correct.
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html#RESTObjectPUT-requests
Look for the section titled "Server-Side-Encryption-Specific Request Headers".  Under "Use customer-provided encryption keys" there's a note saying:

> If you use this feature, the ETag value that Amazon S3 returns in the response is not the MD5 of the object.

There is no such note under the section talking about using an AWS managed key.

How unfortunate.

**Comment by frittentheke** [comment]
While this is not an immediate response the actual issue ... how about moving away from having the user use etag value at all to recognize that the file has changed and use an arbitrary checksum of the local file and store it in the terraform state after a successful upload? This way doing the checksum on the source and comparing it to the terraform state should be enough to recognize a changed file and the requirement to update the S3 object. Using the etag is unsafe in multiple cases - https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html

Or am I missing anything / a certain use case there?

**Comment by eedwards-sk** [comment]
any updates on this? @frittentheke makes a great suggestion

I'm currently bit by needing server side encryption but then being unable to have terraform manage the object due to inability do recognize changes.

**Comment by frittentheke** [comment]
@nwalke @marek-obuchowicz @eedwards-sk @akkessler @mikemac42 as you see referenced, I opened a new issue just about the independently done hash approach of tracking file changes - see https://github.com/terraform-providers/terraform-provider-aws/issues/6668

**Comment by bflad** [comment]
Support for specifying the `aws_s3_bucket_object` resource `etag` argument with `server_side_encryption` (only supporting SSE-S3) has been merged and will release with version 2.25.0 of the Terraform AWS Provider later this week. The resource will currently not provide any error if you attempt to use other forms of SSE with `etag`, but that can be implemented in a separate change request.

The feature request for a separate hash/checksum argument in #6668 is certainly still valid to cover the usage of other forms of SSE. Please follow that issue for updates there. Thanks.
